Description: Restore GNOME2-like Toolbars
 Modified kiryb33's patch to work with SolusOS Nautilus, i.e. no symbolic icons
 The original patch added new actions to the toolbar, previously seen in older
 Gnome 2.30/2.32 Nautilus versions. We've also moved the toolbar actions to the
 start as opposed to the end, and made the bar look that lil' bit more cohesive.
 .
 nautilus (1:3.4.2-1.3) two; urgency=low
 .
   * Better toolbars, modified version of kirby33's patch, emblems dependency by default
Author: Ikey Doherty <ikey@solusos.com>

---
Origin: other, https://launchpadlibrarian.net/108318716/nautiluspatch_3.4.2-0ubuntu3-2.txt
Reviewed-By: Ikey Doherty <ikey@solusos.com>
Last-Update: <2012-07-07>

--- nautilus-3.4.2.orig/src/nautilus-tree-sidebar.c
+++ nautilus-3.4.2/src/nautilus-tree-sidebar.c
@@ -1333,7 +1333,12 @@ create_tree (FMTreeView *view)
 		(view->details->child_model, "row_loaded",
 		 G_CALLBACK (row_loaded_callback),
 		 view, G_CONNECT_AFTER);
-	home_uri = nautilus_get_home_directory_uri ();
+#ifdef NOT_YET_USABLE /* Do we really want this? */
+	icon = g_themed_icon_new (NAUTILUS_ICON_COMPUTER);
+	fm_tree_model_add_root_uri (view->details->child_model, "computer:///", _("Computer"), icon, NULL);
+	g_object_unref (icon);
+#endif
+ 	home_uri = nautilus_get_home_directory_uri ();
 	icon = g_themed_icon_new (NAUTILUS_ICON_HOME);
 	fm_tree_model_add_root_uri (view->details->child_model, home_uri, _("Home"), icon, NULL);
 	g_object_unref (icon);
@@ -1343,7 +1348,7 @@ create_tree (FMTreeView *view)
 	g_object_unref (icon);
 #ifdef NOT_YET_USABLE /* Do we really want this? */
 	icon = g_themed_icon_new (NAUTILUS_ICON_NETWORK);
-	fm_tree_model_add_root_uri (view->details->child_model, "network:///", _("Network Neighbourhood"), icon, NULL);
+	fm_tree_model_add_root_uri (view->details->child_model, "network:///", _("Network"), icon, NULL);
 	g_object_unref (icon);
 #endif
 	
--- nautilus-3.4.2.orig/src/nautilus-toolbar.c
+++ nautilus-3.4.2/src/nautilus-toolbar.c
@@ -64,6 +64,10 @@ G_DEFINE_TYPE (NautilusToolbar, nautilus
 static void
 toolbar_update_appearance (NautilusToolbar *self)
 {
+	GtkAction *action;
+	GtkWidget *widgetitem;
+	gboolean icon_toolbar;
+
 	gboolean show_location_entry;
 
 	show_location_entry = self->priv->show_location_entry ||
@@ -79,6 +83,37 @@ toolbar_update_appearance (NautilusToolb
 
 	gtk_widget_set_visible (self->priv->search_bar,
 				self->priv->show_search_bar);
+
+
+	widgetitem = gtk_ui_manager_get_widget (self->priv->ui_manager, "/Toolbar/Up");
+	icon_toolbar = g_settings_get_boolean (nautilus_preferences, NAUTILUS_PREFERENCES_SHOW_UP_ICON_TOOLBAR);
+	if ( icon_toolbar == FALSE ) { gtk_widget_hide (widgetitem); }
+	else {gtk_widget_show (widgetitem);}
+
+	widgetitem = gtk_ui_manager_get_widget (self->priv->ui_manager, "/Toolbar/Reload");
+	icon_toolbar = g_settings_get_boolean (nautilus_preferences, NAUTILUS_PREFERENCES_SHOW_RELOAD_ICON_TOOLBAR);
+	if ( icon_toolbar == FALSE ) { gtk_widget_hide (widgetitem); }
+	else {gtk_widget_show (widgetitem);}
+
+	widgetitem = gtk_ui_manager_get_widget (self->priv->ui_manager, "/Toolbar/Edit");
+	icon_toolbar = g_settings_get_boolean (nautilus_preferences, NAUTILUS_PREFERENCES_SHOW_EDIT_ICON_TOOLBAR);
+	if ( icon_toolbar == FALSE ) { gtk_widget_hide (widgetitem); }
+	else {gtk_widget_show (widgetitem);}
+
+	widgetitem = gtk_ui_manager_get_widget (self->priv->ui_manager, "/Toolbar/Home");
+	icon_toolbar = g_settings_get_boolean (nautilus_preferences, NAUTILUS_PREFERENCES_SHOW_HOME_ICON_TOOLBAR);
+	if ( icon_toolbar == FALSE ) { gtk_widget_hide (widgetitem); }
+	else {gtk_widget_show (widgetitem);}
+
+	widgetitem = gtk_ui_manager_get_widget (self->priv->ui_manager, "/Toolbar/Computer");
+	icon_toolbar = g_settings_get_boolean (nautilus_preferences, NAUTILUS_PREFERENCES_SHOW_COMPUTER_ICON_TOOLBAR);
+	if ( icon_toolbar == FALSE ) { gtk_widget_hide (widgetitem); }
+	else {gtk_widget_show (widgetitem);}
+
+	widgetitem = gtk_ui_manager_get_widget (self->priv->ui_manager, "/Toolbar/Search");
+	icon_toolbar = g_settings_get_boolean (nautilus_preferences, NAUTILUS_PREFERENCES_SHOW_SEARCH_ICON_TOOLBAR);
+	if ( icon_toolbar == FALSE ) { gtk_widget_hide (widgetitem); }
+	else {gtk_widget_show (widgetitem);}
 }
 
 static void
@@ -101,7 +136,8 @@ nautilus_toolbar_constructed (GObject *o
 
 	toolbar = gtk_ui_manager_get_widget (self->priv->ui_manager, "/Toolbar");
 	self->priv->toolbar = toolbar;
-	gtk_toolbar_set_icon_size (GTK_TOOLBAR (toolbar), GTK_ICON_SIZE_SMALL_TOOLBAR);
+	/** Remove this, should be up to the user what the icon size is.
+	gtk_toolbar_set_icon_size (GTK_TOOLBAR (toolbar), GTK_ICON_SIZE_SMALL_TOOLBAR); */
 
 	context = gtk_widget_get_style_context (toolbar);
 	gtk_style_context_add_class (context, GTK_STYLE_CLASS_PRIMARY_TOOLBAR);
@@ -127,7 +163,8 @@ nautilus_toolbar_constructed (GObject *o
 	item = gtk_tool_item_new ();
 	gtk_tool_item_set_expand (item, TRUE);
 	gtk_container_add (GTK_CONTAINER (item), hbox);
-	gtk_toolbar_insert (GTK_TOOLBAR (self->priv->toolbar), item, 0);
+	/* append to the end of the toolbar so navigation buttons are at the beginning */
+	gtk_toolbar_insert (GTK_TOOLBAR (self->priv->toolbar), item, -1);
 	gtk_widget_show (GTK_WIDGET (item));
 
 	/* search bar */
@@ -138,7 +175,31 @@ nautilus_toolbar_constructed (GObject *o
 				  "changed::" NAUTILUS_PREFERENCES_ALWAYS_USE_LOCATION_ENTRY,
 				  G_CALLBACK (toolbar_update_appearance), self);
 
+	/* nautilus patch */
+	g_signal_connect_swapped (nautilus_preferences,
+				  "changed::" NAUTILUS_PREFERENCES_SHOW_UP_ICON_TOOLBAR,
+				  G_CALLBACK (toolbar_update_appearance), self);
+	g_signal_connect_swapped (nautilus_preferences,
+				  "changed::" NAUTILUS_PREFERENCES_SHOW_EDIT_ICON_TOOLBAR,
+				  G_CALLBACK (toolbar_update_appearance), self);
+	g_signal_connect_swapped (nautilus_preferences,
+				  "changed::" NAUTILUS_PREFERENCES_SHOW_RELOAD_ICON_TOOLBAR,
+				  G_CALLBACK (toolbar_update_appearance), self);
+	g_signal_connect_swapped (nautilus_preferences,
+				  "changed::" NAUTILUS_PREFERENCES_SHOW_HOME_ICON_TOOLBAR,
+				  G_CALLBACK (toolbar_update_appearance), self);
+	g_signal_connect_swapped (nautilus_preferences,
+				  "changed::" NAUTILUS_PREFERENCES_SHOW_COMPUTER_ICON_TOOLBAR,
+				  G_CALLBACK (toolbar_update_appearance), self);
+	g_signal_connect_swapped (nautilus_preferences,
+				  "changed::" NAUTILUS_PREFERENCES_SHOW_SEARCH_ICON_TOOLBAR,
+				  G_CALLBACK (toolbar_update_appearance), self);
+	g_signal_connect_swapped (nautilus_preferences,
+				  "changed::" NAUTILUS_PREFERENCES_SHOW_LABEL_SEARCH_ICON_TOOLBAR,
+				  G_CALLBACK (toolbar_update_appearance), self);
+
 	toolbar_update_appearance (self);
+
 }
 
 static void
@@ -282,6 +343,12 @@ nautilus_toolbar_get_search_bar (Nautilu
 	return self->priv->search_bar;
 }
 
+GtkWidget *
+nautilus_toolbar_get_show_location_entry (NautilusToolbar *self)
+{
+	return self->priv->show_location_entry;
+}
+
 void
 nautilus_toolbar_set_show_main_bar (NautilusToolbar *self,
 				    gboolean show_main_bar)
--- nautilus-3.4.2.orig/src/nautilus-actions.h
+++ nautilus-3.4.2/src/nautilus-actions.h
@@ -29,6 +29,7 @@
 #define NAUTILUS_ACTION_STOP "Stop"
 #define NAUTILUS_ACTION_RELOAD "Reload"
 #define NAUTILUS_ACTION_BACK "Back"
+#define NAUTILUS_ACTION_COMPUTER "Computer"
 #define NAUTILUS_ACTION_UP "Up"
 #define NAUTILUS_ACTION_UP_ACCEL "UpAccel"
 #define NAUTILUS_ACTION_UP_ACCEL "UpAccel"
@@ -43,6 +44,7 @@
 #define NAUTILUS_ACTION_GO_HOME "Home"
 #define NAUTILUS_ACTION_ADD_BOOKMARK "Add Bookmark"
 #define NAUTILUS_ACTION_EDIT_BOOKMARKS "Edit Bookmarks"
+#define NAUTILUS_ACTION_EDIT "Edit"
 #define NAUTILUS_ACTION_HOME "Home"
 #define NAUTILUS_ACTION_ZOOM_IN "Zoom In"
 #define NAUTILUS_ACTION_ZOOM_OUT "Zoom Out"
--- nautilus-3.4.2.orig/src/nautilus-window.c
+++ nautilus-3.4.2/src/nautilus-window.c
@@ -2055,6 +2055,7 @@ nautilus_window_split_view_on (NautilusW
 void
 nautilus_window_split_view_off (NautilusWindow *window)
 {
+	gboolean show_label_search_icon_toolbar;
 	NautilusWindowPane *pane, *active_pane;
 	GList *l, *next;
 
@@ -2074,7 +2075,9 @@ nautilus_window_split_view_off (Nautilus
 					      active_pane->action_group);
 
 	nautilus_window_update_show_hide_menu_items (window);
-	window_set_search_action_text (window, TRUE);
+
+	show_label_search_icon_toolbar = g_settings_get_boolean (nautilus_preferences, NAUTILUS_PREFERENCES_SHOW_LABEL_SEARCH_ICON_TOOLBAR);
+	window_set_search_action_text (window, show_label_search_icon_toolbar);
 }
 
 gboolean
--- nautilus-3.4.2.orig/src/nautilus-toolbar.h
+++ nautilus-3.4.2/src/nautilus-toolbar.h
@@ -64,6 +64,7 @@ GType nautilus_toolbar_get_type (void);
 
 GtkWidget *nautilus_toolbar_new (GtkActionGroup *action_group);
 
+GtkWidget *nautilus_toolbar_get_show_location_entry (NautilusToolbar *self);
 GtkWidget *nautilus_toolbar_get_path_bar (NautilusToolbar *self);
 GtkWidget *nautilus_toolbar_get_location_bar (NautilusToolbar *self);
 GtkWidget *nautilus_toolbar_get_search_bar (NautilusToolbar *self);
--- nautilus-3.4.2.orig/src/nautilus-window-menus.c
+++ nautilus-3.4.2/src/nautilus-window-menus.c
@@ -316,6 +316,7 @@ action_about_nautilus_callback (GtkActio
 		"Andy Hertzfeld",
 		"Arlo Rose",
 		"Christian Neumair",
+  		"Christian Raya",
 		"Cosimo Cecchi",
 		"Darin Adler",
 		"David Camp",
@@ -1029,8 +1030,8 @@ static const GtkActionEntry main_entries
   /* name, stock id, label */  { NAUTILUS_ACTION_FORWARD, GTK_STOCK_GO_FORWARD, N_("_Forward"),
 				 "<alt>Right", N_("Go to the next visited location"),
 				 G_CALLBACK (action_forward_callback) },
-  /* name, stock id, label */  { "Go to Location", NULL, N_("_Location..."),
-                                 "<control>L", N_("Specify a location to open"),
+  /* name, stock id, label */  { "Go to Location", NULL, N_("Toggle _Location bar / Path bar"),
+                                 "F12", N_("Specify a location to open"),
                                  G_CALLBACK (action_go_to_location_callback) },
   /* name, stock id, label */  { "SplitViewNextPane", NULL, N_("S_witch to Other Pane"),
 				 "F6", N_("Move focus to the other pane in a split view window"),
@@ -1104,6 +1105,8 @@ static const GtkRadioActionEntry main_ra
 GtkActionGroup *
 nautilus_window_create_toolbar_action_group (NautilusWindow *window)
 {
+	gboolean show_label_search_icon_toolbar;
+
 	NautilusNavigationState *navigation_state;
 	GtkActionGroup *action_group;
 	GtkAction *action;
@@ -1143,16 +1146,92 @@ nautilus_window_create_toolbar_action_gr
 
 	g_object_unref (action);
 
-	action = GTK_ACTION
-		(gtk_toggle_action_new (NAUTILUS_ACTION_SEARCH,
-					_("Search"),
-					_("Search documents and folders by name"),
-					NULL));
-	gtk_action_group_add_action (action_group, action);
-	gtk_action_set_icon_name (GTK_ACTION (action), "edit-find");
-	gtk_action_set_is_important (GTK_ACTION (action), TRUE);
-
-	g_object_unref (action);
+	/**
+	 * Nautilus 2.30/2.32 type actions
+	 */
+   	action = g_object_new (NAUTILUS_TYPE_NAVIGATION_ACTION,
+   			       "name", NAUTILUS_ACTION_UP,
+   			       "label", _("_Up"),
+   			       "stock_id", GTK_STOCK_GO_UP,
+   			       "tooltip", _("Go to parent folder"),
+   			       "arrow-tooltip", _("Forward history"),
+   			       "window", window,
+   			       "direction", NAUTILUS_NAVIGATION_DIRECTION_UP,
+   			       NULL);
+   	g_signal_connect (action, "activate",
+   			  G_CALLBACK (action_up_callback), window);
+   	gtk_action_group_add_action (action_group, action);
+   
+   	g_object_unref (action);
+  
+   	action = g_object_new (NAUTILUS_TYPE_NAVIGATION_ACTION,
+   			       "name", NAUTILUS_ACTION_RELOAD,
+   			       "label", _("_Reload"),
+   			       "stock_id", GTK_STOCK_REFRESH,
+   			       "tooltip", _("Reload the current location"),
+   			       "window", window,
+   			       "direction", NAUTILUS_NAVIGATION_DIRECTION_RELOAD,
+   			       NULL);
+   	g_signal_connect (action, "activate",
+   			  G_CALLBACK (action_reload_callback), window);
+   	gtk_action_group_add_action (action_group, action);
+   	
+   	g_object_unref (action);
+   
+   	action = g_object_new (NAUTILUS_TYPE_NAVIGATION_ACTION,
+   			       "name", NAUTILUS_ACTION_HOME,
+   			       "label", _("_Home"),
+   			       "stock_id", GTK_STOCK_HOME,
+   			       "tooltip", _("Go to home directory"),
+   			       "window", window,
+   			       "direction", NAUTILUS_NAVIGATION_DIRECTION_HOME,
+   			       NULL);
+   	g_signal_connect (action, "activate",
+   			  G_CALLBACK (action_home_callback), window);
+   	gtk_action_group_add_action (action_group, action);
+   
+   	g_object_unref (action);
+   
+   	action = g_object_new (NAUTILUS_TYPE_NAVIGATION_ACTION,
+   			       "name", NAUTILUS_ACTION_COMPUTER,
+   			       "label", _("_Computer"),
+   			       "stock_id", GTK_STOCK_HARDDISK,
+   			       "tooltip", _("Go to Computer"),
+   			       "window", window,
+   			       "direction", NAUTILUS_NAVIGATION_DIRECTION_COMPUTER,
+   			       NULL);
+   	g_signal_connect (action, "activate",
+   			  G_CALLBACK (action_go_to_computer_callback), window);
+   	gtk_action_group_add_action (action_group, action);
+   
+   	g_object_unref (action);
+ 
+   	action = g_object_new (NAUTILUS_TYPE_NAVIGATION_ACTION,
+   			       "name", NAUTILUS_ACTION_EDIT,
+   			       "label", _("Location"),
+   			       "stock_id", GTK_STOCK_EDIT,
+   			       "tooltip", _("Toggle Location bar / Path bar"),
+   			       "window", window,
+    			       "direction", NAUTILUS_NAVIGATION_DIRECTION_EDIT,
+   			       NULL);
+   	g_signal_connect (action, "activate",
+   			  G_CALLBACK (action_go_to_location_callback), window);
+   	gtk_action_group_add_action (action_group, action);
+  
+   	g_object_unref (action);
+
+ 	action = GTK_ACTION (gtk_toggle_action_new (NAUTILUS_ACTION_SEARCH,
+ 				_("Search"),_("Search documents and folders by name"),
+ 				NULL));
+ 
+  	gtk_action_group_add_action (action_group, action);
+  	gtk_action_set_icon_name (GTK_ACTION (action), "edit-find");
+ 
+ 
+ 	show_label_search_icon_toolbar = g_settings_get_boolean (nautilus_preferences, NAUTILUS_PREFERENCES_SHOW_LABEL_SEARCH_ICON_TOOLBAR);
+ 	gtk_action_set_is_important (GTK_ACTION (action), show_label_search_icon_toolbar);
+  
+  	g_object_unref (action);
 
 	navigation_state = nautilus_window_get_navigation_state (window);
 	nautilus_navigation_state_add_group (navigation_state, action_group);
@@ -1201,7 +1280,7 @@ nautilus_window_initialize_actions (Naut
 {
 	GtkActionGroup *action_group;
 	const gchar *nav_state_actions[] = {
-		NAUTILUS_ACTION_BACK, NAUTILUS_ACTION_FORWARD,
+		NAUTILUS_ACTION_BACK, NAUTILUS_ACTION_FORWARD, NAUTILUS_ACTION_UP, NAUTILUS_ACTION_RELOAD, NAUTILUS_ACTION_COMPUTER, NAUTILUS_ACTION_HOME, NAUTILUS_ACTION_EDIT,
 		NAUTILUS_ACTION_SEARCH, NULL
 	};
 
@@ -1255,6 +1334,9 @@ nautilus_window_initialize_menus (Nautil
 	action = gtk_action_group_get_action (action_group, NAUTILUS_ACTION_HOME);
 	g_object_set (action, "short_label", _("_Home"), NULL);
 
+  	action = gtk_action_group_get_action (action_group, NAUTILUS_ACTION_EDIT);
+  	g_object_set (action, "short_label", _("_Location"), NULL);
+
 	action = gtk_action_group_get_action (action_group, NAUTILUS_ACTION_SHOW_HIDDEN_FILES);
 	g_signal_handlers_block_by_func (action, action_show_hidden_files_callback, window);
 	gtk_toggle_action_set_active (GTK_TOGGLE_ACTION (action),
--- nautilus-3.4.2.orig/src/nautilus-toolbar-ui.xml
+++ nautilus-3.4.2/src/nautilus-toolbar-ui.xml
@@ -2,6 +2,11 @@
 <toolbar name="Toolbar">
 	<toolitem name="Back" action="Back"/>
 	<toolitem name="Forward" action="Forward"/>
+        <toolitem name="Up" action="Up"/>
+ 	<toolitem name="Reload" action="Reload"/>
+        <toolitem name="Edit" action="Edit"/>
+        <toolitem name="Home" action="Home"/>
+        <toolitem name="Computer" action="Computer"/>
         <toolitem name="Search" action="Search"/>
 </toolbar>
-</ui>
\ No newline at end of file
+</ui>
--- nautilus-3.4.2.orig/src/nautilus-window-pane.c
+++ nautilus-3.4.2/src/nautilus-window-pane.c
@@ -1048,10 +1048,17 @@ nautilus_window_pane_grab_focus (Nautilu
 void
 nautilus_window_pane_ensure_location_bar (NautilusWindowPane *pane)
 {
+ 	gboolean fl_active;
 	remember_focus_widget (pane);
 
 	nautilus_toolbar_set_show_main_bar (NAUTILUS_TOOLBAR (pane->tool_bar), TRUE);
-	nautilus_toolbar_set_show_location_entry (NAUTILUS_TOOLBAR (pane->tool_bar), TRUE);
+  
+  	fl_active = nautilus_toolbar_get_show_location_entry (NAUTILUS_TOOLBAR (pane->tool_bar));
+  	if (fl_active) {
+  		nautilus_toolbar_set_show_location_entry (NAUTILUS_TOOLBAR (pane->tool_bar), FALSE);
+ 	} else {
+ 		nautilus_toolbar_set_show_location_entry (NAUTILUS_TOOLBAR (pane->tool_bar), TRUE);
+ 	}
 
 	if (!g_settings_get_boolean (nautilus_window_state,
 				     NAUTILUS_WINDOW_STATE_START_WITH_TOOLBAR)) {
--- nautilus-3.4.2.orig/src/nautilus-file-management-properties.c
+++ nautilus-3.4.2/src/nautilus-file-management-properties.c
@@ -62,6 +62,15 @@
 #define NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_HIDDEN_WIDGET "hidden_files_checkbutton"
 #define NAUTILUS_FILE_MANAGEMENT_PROPERTIES_TREE_VIEW_FOLDERS_WIDGET "treeview_folders_checkbutton"
 
+#define NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_UP_ICON_TOOLBAR_WIDGET "show_up_icon_toolbar_checkbutton"
+#define NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_RELOAD_ICON_TOOLBAR_WIDGET "show_reload_icon_toolbar_checkbutton"
+#define NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_EDIT_ICON_TOOLBAR_WIDGET "show_edit_icon_toolbar_checkbutton"
+#define NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_HOME_ICON_TOOLBAR_WIDGET "show_home_icon_toolbar_checkbutton"
+#define NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_COMPUTER_ICON_TOOLBAR_WIDGET "show_computer_icon_toolbar_checkbutton"
+#define NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_SEARCH_ICON_TOOLBAR_WIDGET "show_search_icon_toolbar_checkbutton"
+#define NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_LABEL_SEARCH_ICON_TOOLBAR_WIDGET "show_label_search_icon_toolbar_checkbutton"
+
+
 /* int enums */
 #define NAUTILUS_FILE_MANAGEMENT_PROPERTIES_THUMBNAIL_LIMIT_WIDGET "preview_image_size_combobox"
 
@@ -720,6 +729,30 @@ nautilus_file_management_properties_dial
 							       4);
 	create_date_format_menu (builder);
 
+
+	/* nautilus patch */
+	bind_builder_bool (builder, nautilus_preferences,
+			   NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_UP_ICON_TOOLBAR_WIDGET,
+			   NAUTILUS_PREFERENCES_SHOW_UP_ICON_TOOLBAR);
+	bind_builder_bool (builder, nautilus_preferences,
+			   NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_RELOAD_ICON_TOOLBAR_WIDGET,
+			   NAUTILUS_PREFERENCES_SHOW_RELOAD_ICON_TOOLBAR);
+	bind_builder_bool (builder, nautilus_preferences,
+			   NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_EDIT_ICON_TOOLBAR_WIDGET,
+			   NAUTILUS_PREFERENCES_SHOW_EDIT_ICON_TOOLBAR);
+	bind_builder_bool (builder, nautilus_preferences,
+			   NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_HOME_ICON_TOOLBAR_WIDGET,
+			   NAUTILUS_PREFERENCES_SHOW_HOME_ICON_TOOLBAR);
+	bind_builder_bool (builder, nautilus_preferences,
+			   NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_COMPUTER_ICON_TOOLBAR_WIDGET,
+			   NAUTILUS_PREFERENCES_SHOW_COMPUTER_ICON_TOOLBAR);
+	bind_builder_bool (builder, nautilus_preferences,
+			   NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_SEARCH_ICON_TOOLBAR_WIDGET,
+			   NAUTILUS_PREFERENCES_SHOW_SEARCH_ICON_TOOLBAR);
+	bind_builder_bool (builder, nautilus_preferences,
+			   NAUTILUS_FILE_MANAGEMENT_PROPERTIES_SHOW_LABEL_SEARCH_ICON_TOOLBAR_WIDGET,
+			   NAUTILUS_PREFERENCES_SHOW_LABEL_SEARCH_ICON_TOOLBAR);
+
 	/* setup preferences */
 	bind_builder_bool (builder, nautilus_icon_view_preferences,
 			   NAUTILUS_FILE_MANAGEMENT_PROPERTIES_LABELS_BESIDE_ICONS_WIDGET,
--- nautilus-3.4.2.orig/src/nautilus-file-management-properties.ui
+++ nautilus-3.4.2/src/nautilus-file-management-properties.ui
@@ -603,7 +603,7 @@
                         <property name="position">1</property>
                       </packing>
                     </child>
-                  </object>
+                   </object>
                   <packing>
                     <property name="expand">False</property>
                     <property name="fill">True</property>
@@ -1607,6 +1607,185 @@
                 <property name="tab_fill">False</property>
               </packing>
             </child>
+            <child>
+              <object class="GtkBox" id="vbox37">
+                <property name="visible">True</property>
+                <property name="can_focus">False</property>
+                <property name="border_width">12</property>
+                <property name="orientation">vertical</property>
+                <property name="spacing">18</property>
+                <child>
+                  <object class="GtkBox" id="vbox38">
+                    <property name="visible">True</property>
+                    <property name="can_focus">False</property>
+                    <property name="orientation">vertical</property>
+                    <property name="spacing">6</property>
+                    <child>
+                      <object class="GtkLabel" id="label39">
+                        <property name="visible">True</property>
+                        <property name="can_focus">False</property>
+                        <property name="xalign">0</property>
+                        <property name="label" translatable="yes">&lt;b&gt;Options&lt;/b&gt;</property>
+                        <property name="use_markup">True</property>
+                      </object>
+                      <packing>
+                        <property name="expand">False</property>
+                        <property name="fill">False</property>
+                        <property name="position">0</property>
+                      </packing>
+                    </child>
+                    <child>
+                      <object class="GtkAlignment" id="alignment22">
+                        <property name="visible">True</property>
+                        <property name="can_focus">False</property>
+                        <property name="left_padding">12</property>
+                        <child>
+                          <object class="GtkBox" id="vbox39">
+                            <property name="visible">True</property>
+                            <property name="can_focus">False</property>
+                            <property name="orientation">vertical</property>
+                            <property name="spacing">6</property>
+                            <child>
+                              <object class="GtkCheckButton" id="show_up_icon_toolbar_checkbutton">
+                                <property name="label" translatable="yes">Show up icon</property>
+                                <property name="visible">True</property>
+                                <property name="can_focus">True</property>
+                                <property name="receives_default">False</property>
+                                <property name="use_action_appearance">False</property>
+                                <property name="use_underline">True</property>
+                                <property name="draw_indicator">True</property>
+                              </object>
+                              <packing>
+                                <property name="expand">False</property>
+                                <property name="fill">False</property>
+                                <property name="position">0</property>
+                              </packing>
+                            </child>
+                            <child>
+                              <object class="GtkCheckButton" id="show_reload_icon_toolbar_checkbutton">
+                                <property name="label" translatable="yes">Show refresh icon</property>
+                                <property name="visible">True</property>
+                                <property name="can_focus">True</property>
+                                <property name="receives_default">False</property>
+                                <property name="use_action_appearance">False</property>
+                                <property name="use_underline">True</property>
+                                <property name="draw_indicator">True</property>
+                              </object>
+                              <packing>
+                                <property name="expand">False</property>
+                                <property name="fill">False</property>
+                                <property name="position">1</property>
+                              </packing>
+                            </child>
+                             <child>
+                              <object class="GtkCheckButton" id="show_edit_icon_toolbar_checkbutton">
+                                <property name="label" translatable="yes">Show toggle button location bar/bar path</property>
+                                <property name="visible">True</property>
+                                <property name="can_focus">True</property>
+                                <property name="receives_default">False</property>
+                                <property name="use_action_appearance">False</property>
+                                <property name="use_underline">True</property>
+                                <property name="draw_indicator">True</property>
+                              </object>
+                              <packing>
+                                <property name="expand">False</property>
+                                <property name="fill">False</property>
+                                <property name="position">2</property>
+                              </packing>
+                            </child>
+                            <child>
+                              <object class="GtkCheckButton" id="show_home_icon_toolbar_checkbutton">
+                                <property name="label" translatable="yes">Show home icon</property>
+                                <property name="visible">True</property>
+                                <property name="can_focus">True</property>
+                                <property name="receives_default">False</property>
+                                <property name="use_action_appearance">False</property>
+                                <property name="use_underline">True</property>
+                                <property name="draw_indicator">True</property>
+                              </object>
+                              <packing>
+                                <property name="expand">False</property>
+                                <property name="fill">False</property>
+                                <property name="position">3</property>
+                              </packing>
+                            </child>
+                            <child>
+                              <object class="GtkCheckButton" id="show_computer_icon_toolbar_checkbutton">
+                                <property name="label" translatable="yes">Show computer icon</property>
+                                <property name="visible">True</property>
+                                <property name="can_focus">True</property>
+                                <property name="receives_default">False</property>
+                                <property name="use_action_appearance">False</property>
+                                <property name="use_underline">True</property>
+                                <property name="draw_indicator">True</property>
+                              </object>
+                              <packing>
+                                <property name="expand">False</property>
+                                <property name="fill">False</property>
+                                <property name="position">4</property>
+                              </packing>
+                            </child>
+                            <child>
+                              <object class="GtkCheckButton" id="show_search_icon_toolbar_checkbutton">
+                                <property name="label" translatable="yes">Show search icon</property>
+                                <property name="visible">True</property>
+                                <property name="can_focus">True</property>
+                                <property name="receives_default">False</property>
+                                <property name="use_action_appearance">False</property>
+                                <property name="use_underline">True</property>
+                                <property name="draw_indicator">True</property>
+                              </object>
+                              <packing>
+                                <property name="expand">False</property>
+                                <property name="fill">False</property>
+                                <property name="position">5</property>
+                              </packing>
+                            </child>
+                            <child>
+                              <object class="GtkCheckButton" id="show_label_search_icon_toolbar_checkbutton">
+                                <property name="label" translatable="yes">Show label on search icon (need to restart nautilus)</property>
+                                <property name="visible">True</property>
+                                <property name="can_focus">True</property>
+                                <property name="receives_default">False</property>
+                                <property name="use_action_appearance">False</property>
+                                <property name="use_underline">True</property>
+                                <property name="draw_indicator">True</property>
+                              </object>
+                              <packing>
+                                <property name="expand">False</property>
+                                <property name="fill">False</property>
+                                <property name="position">6</property>
+                              </packing>
+                            </child>
+                          </object>
+                        </child>
+                      </object>
+                      <packing>
+                        <property name="expand">False</property>
+                        <property name="fill">True</property>
+                        <property name="position">1</property>
+                      </packing>
+                    </child>
+                  </object>
+                  <packing>
+                    <property name="expand">False</property>
+                    <property name="fill">True</property>
+                    <property name="position">5</property>
+                  </packing>
+                </child>
+              </object>
+            </child>
+            <child type="tab">
+              <object class="GtkLabel" id="label38">
+                <property name="visible">True</property>
+                <property name="can_focus">False</property>
+                <property name="label" translatable="yes">Toolbar</property>
+              </object>
+              <packing>
+                <property name="position">2</property>
+                <property name="tab_fill">False</property>
+              </packing>
+            </child>
           </object>
           <packing>
             <property name="expand">False</property>
--- nautilus-3.4.2.orig/src/nautilus-navigation-action.c
+++ nautilus-3.4.2/src/nautilus-navigation-action.c
@@ -149,6 +149,16 @@ show_menu (NautilusNavigationAction *sel
 	case NAUTILUS_NAVIGATION_DIRECTION_BACK:
 		fill_menu (window, menu, TRUE);
 		break;
+ 	case NAUTILUS_NAVIGATION_DIRECTION_UP:
+ 		return;
+ 	case NAUTILUS_NAVIGATION_DIRECTION_RELOAD:
+ 		return;
+ 	case NAUTILUS_NAVIGATION_DIRECTION_HOME:
+ 		return;
+ 	case NAUTILUS_NAVIGATION_DIRECTION_COMPUTER:
+ 		return;
+ 	case NAUTILUS_NAVIGATION_DIRECTION_EDIT:
+ 		return;
 	default:
 		g_assert_not_reached ();
 		break;
--- nautilus-3.4.2.orig/src/nautilus-navigation-action.h
+++ nautilus-3.4.2/src/nautilus-navigation-action.h
@@ -47,7 +47,13 @@ typedef struct NautilusNavigationActionP
 typedef enum
 {
 	NAUTILUS_NAVIGATION_DIRECTION_BACK,
-	NAUTILUS_NAVIGATION_DIRECTION_FORWARD
+   	NAUTILUS_NAVIGATION_DIRECTION_FORWARD,
+   	NAUTILUS_NAVIGATION_DIRECTION_UP,
+   	NAUTILUS_NAVIGATION_DIRECTION_RELOAD,
+   	NAUTILUS_NAVIGATION_DIRECTION_HOME,
+   	NAUTILUS_NAVIGATION_DIRECTION_COMPUTER,
+   	NAUTILUS_NAVIGATION_DIRECTION_EDIT
+ 
 } NautilusNavigationDirection;
 
 struct _NautilusNavigationAction
--- nautilus-3.4.2.orig/libnautilus-private/org.gnome.nautilus.gschema.xml.in
+++ nautilus-3.4.2/libnautilus-private/org.gnome.nautilus.gschema.xml.in
@@ -82,6 +82,41 @@
       <_summary>Always use the location entry, instead of the pathbar</_summary>
       <_description>If set to true, then Nautilus browser windows will always use a textual input entry for the location toolbar, instead of the pathbar.</_description>
     </key>
+    <key type="b" name="show-up-icon-toolbar">
+      <default>false</default>
+      <summary>Show Up button in nautilus toolbar</summary>
+      <description>If set to true, then Nautilus browser windows will show the button.</description>
+    </key>
+    <key type="b" name="show-reload-icon-toolbar">
+      <default>false</default>
+      <summary>Show refresh button in nautilus toolbar</summary>
+      <description>If set to true, then Nautilus browser windows will show the button.</description>
+    </key>
+    <key type="b" name="show-edit-icon-toolbar">
+      <default>false</default>
+      <summary>Show toggle button location entry/pathbar</summary>
+      <description>If set to true, then Nautilus browser windows will show the button.</description>
+    </key>
+    <key type="b" name="show-home-icon-toolbar">
+      <default>false</default>
+      <summary>Show Home button in nautilus toolbar</summary>
+      <description>If set to true, then Nautilus browser windows will show the button.</description>
+    </key>
+    <key type="b" name="show-computer-icon-toolbar">
+      <default>false</default>
+      <summary>Show Computer button in nautilus toolbar</summary>
+      <description>If set to true, then Nautilus browser windows will show the button.</description>
+    </key>
+    <key type="b" name="show-search-icon-toolbar">
+      <default>true</default>
+      <summary>Show Search button in nautilus toolbar</summary>
+      <description>If set to true, then Nautilus browser windows will show the button.</description>
+    </key>
+    <key type="b" name="show-label-search-icon-toolbar">
+      <default>true</default>
+      <summary>Show the Search button label in nautilus toolbar</summary>
+      <description>If set to true, then Nautilus browser windows will show the Search label button.</description>
+    </key>
     <key name="confirm-trash" type="b">
       <default>true</default>
       <_summary>Whether to ask for confirmation when deleting files, or emptying Trash</_summary>
--- nautilus-3.4.2.orig/libnautilus-private/nautilus-global-preferences.h
+++ nautilus-3.4.2/libnautilus-private/nautilus-global-preferences.h
@@ -78,6 +78,13 @@ typedef enum
 #define NAUTILUS_PREFERENCES_NEW_TAB_POSITION			"tabs-open-position"
 
 #define NAUTILUS_PREFERENCES_ALWAYS_USE_LOCATION_ENTRY		"always-use-location-entry"
+#define NAUTILUS_PREFERENCES_SHOW_UP_ICON_TOOLBAR		"show-up-icon-toolbar"
+#define NAUTILUS_PREFERENCES_SHOW_EDIT_ICON_TOOLBAR		"show-edit-icon-toolbar"
+#define NAUTILUS_PREFERENCES_SHOW_RELOAD_ICON_TOOLBAR		"show-reload-icon-toolbar"
+#define NAUTILUS_PREFERENCES_SHOW_HOME_ICON_TOOLBAR		"show-home-icon-toolbar"
+#define NAUTILUS_PREFERENCES_SHOW_COMPUTER_ICON_TOOLBAR		"show-computer-icon-toolbar"
+#define NAUTILUS_PREFERENCES_SHOW_SEARCH_ICON_TOOLBAR		"show-search-icon-toolbar"
+#define NAUTILUS_PREFERENCES_SHOW_LABEL_SEARCH_ICON_TOOLBAR	"show-label-search-icon-toolbar"
 
 /* Which views should be displayed for new windows */
 #define NAUTILUS_WINDOW_STATE_START_WITH_STATUS_BAR		"start-with-status-bar"
--- nautilus-3.4.2.orig/libnautilus-private/nautilus.convert
+++ nautilus-3.4.2/libnautilus-private/nautilus.convert
@@ -2,6 +2,13 @@
 tabs-open-position = /desktop/gnome/file_views/tabs_open_position
 always-use-browser = /apps/nautilus/preferences/always_use_browser
 always-use-location-entry = /apps/nautilus/preferences/always_use_location_entry
+show-up-icon-toolbar = /apps/nautilus/preferences/show-up-icon-toolbar
+show-edit-icon-toolbar = /apps/nautilus/preferences/show-edit-icon-toolbar
+show-reload-icon-toolbar = /apps/nautilus/preferences/show-reload-icon-toolbar
+show-home-icon-toolbar = /apps/nautilus/preferences/show-home-icon-toolbar
+show-computer-icon-toolbar = /apps/nautilus/preferences/show-computer-icon-toolbar
+show-search-icon-toolbar = /apps/nautilus/preferences/show-search-icon-toolbar
+show-label-search-icon-toolbar = /apps/nautilus/preferences/show-label-search-icon-toolbar
 confirm-trash = /apps/nautilus/preferences/confirm_trash
 enable-delete = /apps/nautilus/preferences/enable_delete
 show-icon-text = /apps/nautilus/preferences/show_icon_text
